File: TypeScript/Data Structures/queue.ts
/**
 * Data Structure: Queue
 * Language: Typescript
 */

class Queue<T> {
    private queueList: T[];

    constructor(){
        this.queueList = Array();
    }

    // Returns True if the queue is empty
    isEmpty(): boolean {
        return this.queueList.length == 0;
    }

    // Returns the number of items in the queue
    length(): number {
        return this.queueList.length;
    }

    // Adds the given item to the queue
    enqueue(item: T) {
        this.queueList.unshift(item)
    }

    // Removes and returns the bfirst item in the queue
    dequeue(){
        if (!this.isEmpty()) {
            return this.queueList.pop();
        } else {
            return "Cannot dequeue from an empty queue"
        }
    }
}


// Implementation of a priority queue
class PriorityQueue<PriorityEntry> {
    // create an empty unbounded
    constructor(
        private qList: PriorityEntry[]
    ){}

    // Returns true if the queue is emmpty
    isEmpty(): boolean {
        return this.qList.length == 0;
    }

    // returns the number of items in the queue
    length(): number {
        return this.qList.length;
    }

    // Adds a given item to the queue
    enqueue(queueItem: PriorityEntry){
        const entry: PriorityEntry = queueItem
        this.qList.push(entry)
    }

    // Removes and returns the first item in the queue
    dequeue() {
        let highest:number = 0
        for (let i = 0; i < this.qList.length; i++) {
            if (this.qList[i].priority < highest){
                highest = this.qList[i].priority
            }
        }
        return this.qList.splice(highest, 1)
    }
}

// Private class for storing the associaing queue items with thier priority
interface PriorityEntry {
    item: string,
    priority: number
}

const list: PriorityEntry[] = [
    {item: "hello", priority: 1},
    {item: "master", priority: 2},
    {item: "look", priority: 3},
]

console.log(list[0].priority)

//  Driver Method
let queue = new Queue()
console.log(queue.isEmpty())
queue.enqueue("Tester one")
queue.enqueue("Tester Two")
queue.enqueue("Tester three")
queue.enqueue("Tester four")
console.log(queue.dequeue())

File: Python/Data Structures/queue.py
"""
    Data Structure: Queue
    Language: Python
"""


class Queue:
    def __init__(self):
        self._queueList = list()

    # Returns true if the queue is empty
    def isEmpty(self):
        return len(self._queueList) == 0

    # Returns the number of items in the queue
    def __len__(self):
        return len(self._queueList)

    # Adds the given item to the queue
    def enqueue(self, item):
        self._queueList.append(item)

    # Removes and returns the first item in the queue
    def dequeue(self):
        if not self.isEmpty():
            return self._queueList.pop()
        else:
            return "Cannot dequeue from an empty queue."


class PriorityQueue:
    """Creates an empty unbounded priority queue"""
    def __init__(self):
        self._queueList = list()

    # Returns True if the queue is empty
    def isEmpty(self):
        return len(self._queueList) == 0

    # Returns the number of items in the queue
    def __len__(self):
        return len(self._queueList)

    # Adds the given item to the queue
    def enqueue(self, _item, priority):
        """Create a new instance of the storage cass and append it to the list"""
        entry = _PriorityQueue(_item, priority)
        self._queueList.append(entry)

    # Remove and returns the first item in the queue
    def dequeue(self):
        if not self.isEmpty():
            """Find the entry with highest priority"""
            highest = self._queueList[0].priority
            for i in range(len(self._queueList)):
                """See if the ith entry contains a higher priority (smaller integer)."""
                if self._queueList[i].priority < highest:
                    highest = self._queueList[i].priority
            # Remove the entry with the highest priority and return them
            entry = self._queueList.pop(highest)
            return entry.item
        else:
            return "Cannot dequeue from an empty queue"


class _PriorityQueue:
    def __init__(self, _item, priority):
        self.item = _item
        self.priority = priority


if __name__ == "__main__":
    Q = PriorityQueue()
    Q.enqueue("purple", 5)
    Q.enqueue("black", 1)
    Q.enqueue("orange", 3)
    Q.enqueue("white", 0)
    Q.enqueue("green", 1)
    Q.enqueue("yellow", 5)
    while not Q.isEmpty():
        item = Q.dequeue()
        print(item)
